rfcs/33-apps_own_scheduling.md

```diff
@@ -0,0 +1,239 @@
+# Feature Name: `apps_own_scheduling`
+
+## Summary
+
+All scheduling information is attached directly to systems, and ultimately owned and configurable by the central app (or hand-written `bevy_ecs` equivalent).
+To ensure that this can be done safely, plugins specify invariants about the ordering of their systems which must be respected.
+Hard ordering constraints (where systems must be both ordered and separated by a hard sync point) are introduced as an assert-style protection to guard against a serious and common form of plugin misconfiguration.
+
+## Motivation
+
+The current plugin model works well in simple cases, but completely breaks down when the central consumer (typically, an `App`) needs to configure when and if their systems run. The main difficulties are:
+
+- 3rd party plugins can have "unfixable" ordering ambiguities or interop issues with other 3rd party plugins
+- plugins can not be added to states, and run criteria cannot be added
+- plugins can not be incorporated into apps and games that use a non-standard control flow architecture
+- end users who care about granular scheduling control for performance optimization have no way to change when systems run
+- worse, plugins can freely insert their own stages, causing sudden and uncontrollable bottlenecks
+- plugins cannot be used by standalone `bevy_ecs` users
+
+See [Bevy #2160](https://github.com/bevyengine/bevy/issues/2160) for more background on this problem.
+
+By moving the ultimate responsibility for scheduling configuration to the central app, users can comfortably combine plugins that were not originally designed to work together correctly and fit the logic into their particular control flow.
+
+## User-facing explanation
+
+**Plugins** are cohesive, drop-in units of functionality that you can add to your Bevy app.
+Plugins can initialize **resources** to store data to, or add **system sets**, used to add coherent batches of logic to your app.
+
+Plugins can be added to your app using the `App::add_plugin` method.
+If you're new to Bevy, third-party plugins added in this way should immediately work without further configuration.
+
+```rust
+use bevy::prelude::*;
+use bevy_hypothetical_tui::prelude::*;
+
+fn main(){
+  App::minimal().add_plugin(TuiPlugin).run();
+}
+```
+
+You can write your own plugins by exporting a function that returns a `Plugin` struct with the desired system sets and resources.
+This can be provided as a standalone function or, more commonly, as a method on a simple `MyPlugin` struct:
+
+```rust
+struct CombatPlugin;
+
+impl IntoPlugin for CombatPlugin {
```

Comment:
Maybe better names for `trait IntoPlugin` and `struct Plugin` would be `trait Plugin` (like you said) and `struct PluginDescription`. So something kind of similar to how systems work (I think^^).

Also `IntoPlugin` consuming it's input is not really a problem you could always `impl IntoPlugin for &MyPlugin` (or even `&mut MyPlugin`) if you wanted (same with `Into<Plugin>`).

Reply:
Personally, I think having to declare a separate plugin type is wholly unnecessary, and plugins should simply export functions that return plugin.

Reply:
> Personally, I think having to declare a separate plugin type is wholly unnecessary, and plugins should simply export functions that return plugin.

This was my initial design. Unfortunately it led to an increase in boilerplate over the existing design: you had to call `.add_plugin(MyPlugin::new())` everywhere. I'll lay out the options in Unresolved Questions.

Reply:
Oh wow, my initial type signature was *totally* off. It really wanted a `self` or `&self` in there so it could be configured properly.

Reply:
I've laid out some of the options here: https://github.com/alice-i-cecile/rfcs/commit/c1cc3d786349f10e6a593eb6ccb8b01b3e5998a0

I think that the "crates just export functions that return `Plugin`" is my favorite model here: it's much less convoluted than the alternatives.

Reply:
That first option listed feels like the nicest, though I feel like the example for configured plugin is overly complicated. If it needs a `bool` for configuration, shouldn't the function just take a `bool` instead of a custom type? The newtype pattern and zero sized types are useful, but I feel like we're overusing them a bit at this point...

Reply:
Fair enough. I'll simplify it down in all three cases.

Reply:
I'm personally not a huge fan of "floating exported functions". We do that for systems because it works so well (and the alternatives are significantly worse), but generally I think we should try to export "actual types" over functions. I have a pretty strong preference for plugins to be a user-defined type that implements Plugin (or _maybe_ an "into" variant of that pattern .. but i honestly still like that less). This is for a variety of reasons:
* **Self-documenting / easy to learn**: types _are_ plugins if they implement Plugin. IDEs will even fill in the blanks for you. All "Plugins" look the same and explicitly declare themselves "plugins". If we use functions this gets way more complicated. If you want to export a plugin in your library: define a public function, call it _something_ (`create_my_plugin()`? `my_plugin()`?), and ensure that the function returns a Plugin type. We need to start talking about "ideal plugin construction patterns" and "naming conventions". Is the function that constructs the Plugin your custom plugin? Or is the arbitrary Plugin-struct state returned by the function your custom Plugin?
* **Future-compatible**: what happens if we later decide we want to add  "plugin lifecycle hooks" like `fn on_load()` and `fn on_unload()`. If we use a trait, users just extend what they have (and get helpful error messages if they're missing). If Plugins are custom user types, it also leaves the doors open to using TypeIds to identify specific plugins. What if we want to allow Plugins to define/register their dependent plugins? We could easily "deduplicate" these "dependency registrations" if we can use TypeIds. With "functions returning plugin structs" we need to increase boilerplate by defining some "custom plugin type identity".

Reply:
> That first option listed feels like the nicest, though I feel like the example for configured plugin is overly complicated. If it needs a bool for configuration, shouldn't the function just take a bool instead of a custom type?

I can't actually do this for the example in question, as it needs a new-type wrapper to store the resource in.

Reply:
Well, upon closer inspection that example would give a type error, as you try to put a MyConfig instead of a bool in a MyConfig. Anyways, I'd personally choose to have this wrapping done internally, which would let the plugin be a simple function that takes a bool.

Reply:
> Self-documenting / easy to learn

I'd argue your primary point of "types that impl Plugin are plugins" doesn't feel quite accurate if we introduce a separate `Plugin` type. You point out a similar dilemma for function-plugins, but I'd argue it exists just as much for separate struct based things.

> Future-compatible

Well, in the case of lifecycle hooks, these can simply be stored in `Plugin` structs. And for the possible deduplication strategy, we can simply make it so plugin-accepting functions accept `impl FnOnce() -> Plugin`, and keep using type ids.